core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/instruction/data/number/NumberBetween.java

```diff
@@ -1,43 +1,93 @@
 package io.metadew.iesi.script.execution.instruction.data.number;
 
-import io.metadew.iesi.data.generation.execution.GenerationObjectExecution;
 import io.metadew.iesi.script.execution.instruction.data.DataInstruction;
 
+import java.math.RoundingMode;
+import java.security.SecureRandom;
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
 import java.text.MessageFormat;
+import java.util.Objects;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-/**
- * @author peter.billen
- */
 public class NumberBetween implements DataInstruction {
     private final String LOWER_BOUND_KEY = "LowerBoundRepresentation";
-
     private final String UPPER_BOUND_KEY = "UpperBoundRepresentation";
+    private final String NUMBER_OF_DECIMALS = "NumberOfDecimalRepresentation";
+    private final SecureRandom secureRandom;
 
-    private final Pattern INPUT_PARAMETER_PATTERN = Pattern
-            .compile("\\s*\"?(?<" + LOWER_BOUND_KEY + ">\\d+)\"?\\s*,\\s*\"?(?<" + UPPER_BOUND_KEY + ">\\d+)\"?\\s*");
-
-    private final GenerationObjectExecution generationObjectExecution;
-
-    public NumberBetween(GenerationObjectExecution generationObjectExecution) {
-        this.generationObjectExecution = generationObjectExecution;
+    public NumberBetween() {
+        secureRandom = new SecureRandom();
     }
 
+    private final Pattern INPUT_PARAMETER_PATTERN = Pattern
+            .compile("^\\s*\"?\\s*(?<" + LOWER_BOUND_KEY + ">-?\\d+(?:\\.\\d+)?)\\s*\"?\\s*," +
+                    "\\s*\"?\\s*(?<" + UPPER_BOUND_KEY + ">-?\\d+(?:\\.\\d+)?)\\s*\"?\\s*" +
+                    "(?:,\\s*\"?\\s*(?<" + NUMBER_OF_DECIMALS + ">-?\\d*(?:\\.\\d+)?)\\s*\"?\\s*)?");
+
     @Override
     public String generateOutput(String parameters) {
         Matcher inputParameterMatcher = INPUT_PARAMETER_PATTERN.matcher(parameters);
-        if (!inputParameterMatcher.find()) {
+
+        if (!inputParameterMatcher.find())
             throw new IllegalArgumentException(MessageFormat.format("Illegal arguments provided to " + this.getKeyword() + ": {0}", parameters));
-        } else {
-            double lowerBound = Double.parseDouble(inputParameterMatcher.group(LOWER_BOUND_KEY));
-            double upperBound = Double.parseDouble(inputParameterMatcher.group(UPPER_BOUND_KEY));
-            return Double.toString(generationObjectExecution.getNumber().between(lowerBound, upperBound));
-        }
+
+        String strNumberOfDecimals = inputParameterMatcher.group(NUMBER_OF_DECIMALS);
+
+        double lowerBound = Double.parseDouble(inputParameterMatcher.group(LOWER_BOUND_KEY));
+        double upperBound = Double.parseDouble(inputParameterMatcher.group(UPPER_BOUND_KEY));
+        double generatedNumber = range(lowerBound, upperBound);
+
+        if (Objects.isNull(strNumberOfDecimals))
+            return Double.toString(generatedNumber);
+
+        if (Pattern.compile("\\.").matcher(strNumberOfDecimals).find())
```

Comment:
Yes, the initial regex capture all 3 arguments as integer or float.

As the 3rd group is optional, if I try to put the condition in the first regex then it could resolve as if there wasn't any 3rd argument.

So I'm checking if the user entered a float number instead of an integer to throw an error.

Reply:
added to the code:

    // As g3 is an optional group, the regex can't fail its match because of a bad input in g3 so we allowed the dot
    // to capture an eventual float number and throw an error later